homeassistant/components/prusalink/coordinator.py

```diff
@@ -47,6 +48,8 @@ async def _async_update_data(self) -> T:
             raise UpdateFailed("Invalid authentication") from None
         except PrusaLinkError as err:
             raise UpdateFailed(str(err)) from err
+        except (TimeoutError, ConnectError) as err:
```

Comment:
How would the library user know to catch httpx exceptions when the library has its own exceptions for other errors. This doesn't make sense. The library should have its own exceptions for all errors.

Reply:
This got discussed in the pyprusalink library as well, see: https://github.com/home-assistant-libs/pyprusalink/pull/94#issuecomment-2161651666.

Sure, a library could handle every exception imaginable (OS errors etc.), and convert to their own. But I don't think that is good practice. I think at least for Python built-in exceptions like OS errors etc. it is quite common to let them bubble up.

Now a http library exception is a bit more on the edge here. But in my books, this is also more a "infrastructure" type of Exception. Sure the library user needs to know what exception to handle, but that is also the case Python/low-level exceptions :man_shrugging:

Reply:
I know it got discussed and it still doesn't make sense.

Reply:
> I know it got discussed and it still doesn't make sense.

Why not?

> How would the library user know to catch httpx exceptions when the library has its own exceptions for other errors. This doesn't make sense. The library should have its own exceptions for all errors.

That isn't an argument. The httpx library has many more exceptions we probably don't handle. The underlying librarys might also change, and suddenly raise new exceptions etc. If you have a library which loads a file, you also don't handle every OS/IO error imaginable.

Java tried once to force users to handle every error with their checked exceptions approach. It died. Clean Code by Robert C. Martin has more background on why (mainly because it violated the open-closed principle, it says). If you have the expectation to a library to handles all errors imaginable, you end up in a similar situation.

In my opinion, a library should only handle exceptions where it can add value. And that is what the library does currently, it handles application specific errors and add value by giving more context/meaning.

Just converting a `ConnectError` to a library specific `ConnectError` doesn't add lot's of value. You can make the argument that abstraction is value. But IMHO, just abstraction in this case is not worth the effort. :man_shrugging: